File: frontend/scripts/tldr/tldr_to_db.py
#!/usr/bin/env python3
"""
Build a SQLite database from TLDR markdown files.

- Scans data/tldr/*/*.md
- Creates SQLite DB at db/all_dbs/tldr-db.db
- Table: url_lookup (WITHOUT ROWID)
- Table: cluster
- Table: overview
"""

import hashlib
import json
import re
import sqlite3
import struct
import yaml
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

BASE_DIR = Path(__file__).parent
DATA_DIR = BASE_DIR.parent.parent / "data" / "tldr"
DB_PATH = BASE_DIR.parent.parent / "db" / "all_dbs" / "tldr-db.db"


def create_full_hash(category: str, last_path: str) -> str:
    """Create a SHA-256 hash from category and name."""
    # Normalize input: remove leading/trailing slashes, lowercase
    category = category.strip().lower()
    last_path = last_path.strip().lower()
    
    # Create unique string
    unique_str = f"{category}/{last_path}"
    
    # Compute SHA-256 hash
    return hashlib.sha256(unique_str.encode("utf-8")).hexdigest()


def get_8_bytes(full_hash: str) -> int:
    """Get the first 8 bytes of the hash as a signed 64-bit integer."""
    # Take first 16 hex chars (8 bytes)
    hex_part = full_hash[:16]
    # Convert to bytes
    bytes_val = bytes.fromhex(hex_part)
    # Unpack as signed 64-bit integer (big-endian)
    return struct.unpack(">q", bytes_val)[0]


def ensure_schema(conn: sqlite3.Connection) -> None:
    cur = conn.cursor()

    # Main table for TLDR pages using consistent hashing
    cur.execute(
        """
        CREATE TABLE IF NOT EXISTS url_lookup (
            url_hash INTEGER PRIMARY KEY,
            url TEXT NOT NULL,
            cluster TEXT NOT NULL,          -- Directory name (e.g., 'git', 'aws')
            name TEXT NOT NULL,             -- Command name (e.g., 'git-commit')
            platform TEXT DEFAULT '',       -- From frontmatter 'category' (e.g., 'common', 'linux')
            title TEXT DEFAULT '',          -- From frontmatter
            description TEXT DEFAULT '',    -- From frontmatter or markdown blockquote
            more_info_url TEXT DEFAULT '',  -- From markdown blockquote
            keywords TEXT DEFAULT '[]',     -- JSON array from frontmatter
            features TEXT DEFAULT '[]',     -- JSON array from frontmatter
            examples TEXT DEFAULT '[]',     -- JSON array of {description, cmd}
            raw_content TEXT DEFAULT '',    -- Full markdown content
            path TEXT DEFAULT ''            -- URL path from frontmatter
        ) WITHOUT ROWID;
        """
    )

    # Index for efficient cluster grouping
    cur.execute("CREATE INDEX IF NOT EXISTS idx_url_lookup_cluster ON url_lookup(cluster);")

    # Cluster table (categories)
    cur.execute(
        """
        CREATE TABLE IF NOT EXISTS cluster (
            name TEXT PRIMARY KEY,
            count INTEGER NOT NULL,
            description TEXT DEFAULT '' -- Kept empty as per previous implementation
        );
        """
    )

    # Overview table
    cur.execute(
        """
        CREATE TABLE IF NOT EXISTS overview (
            id INTEGER PRIMARY KEY CHECK(id = 1),
            total_count INTEGER NOT NULL
        );
        """
    )

    conn.commit()


def parse_tldr_file(file_path: Path) -> Optional[Dict[str, Any]]:
    """
    Parse a TLDR markdown file.
    Returns a dictionary with extracted data or None if parsing fails.
    """
    try:
        content = file_path.read_text(encoding="utf-8")
    except Exception as e:
        print(f"Error reading {file_path}: {e}")
        return None

    # Split frontmatter and content
    parts = re.split(r"^---\s*$", content, maxsplit=2, flags=re.MULTILINE)
    if len(parts) < 3:
        print(f"Skipping {file_path.name}: Invalid frontmatter format")
        return None

    frontmatter_raw = parts[1]
    markdown_body = parts[2]

    # Parse Frontmatter
    try:
        fm = yaml.safe_load(frontmatter_raw) or {}
    except yaml.YAMLError as e:
        print(f"Error parsing YAML in {file_path.name}: {e}")
        return None

    # Extract fields from frontmatter
    title = fm.get("title", "")
    # Clean title: remove " | Online Free DevTools by Hexmos" suffix if present
    title = title.split(" | ")[0]
    
    platform = fm.get("category", "")
    path_url = fm.get("path", "")
    keywords = json.dumps(fm.get("keywords", []))
    features = json.dumps(fm.get("features", []))
    
    # Parse Markdown Body
    description_lines = []
    more_info_url = ""
    examples = []
    
    lines = markdown_body.strip().splitlines()
    i = 0
    while i < len(lines):
        line = lines[i].strip()
        
        # Blockquotes (Description & More Info)
        if line.startswith(">"):
            clean_line = line.lstrip("> ").strip()
            if clean_line.startswith("More information:"):
                # Extract URL
                match = re.search(r"<(.*?)>", clean_line)
                if match:
                    more_info_url = match.group(1)
            else:
                description_lines.append(clean_line)
        
        # Examples
        elif line.startswith("- "):
            example_desc = line.lstrip("- ").strip()
            # Look ahead for the command
            cmd = ""
            j = i + 1
            while j < len(lines):
                next_line = lines[j].strip()
                if next_line.startswith("`") and next_line.endswith("`"):
                    cmd = next_line.strip("`")
                    # Handle {{ }} placeholders if needed, but keeping them as is is usually fine for display
                    i = j # Advance main loop
                    break
                elif next_line == "":
                    j += 1
                    continue
                else:
                    # Found something else, stop looking for command
                    break
            
            if cmd:
                examples.append({"description": example_desc, "cmd": cmd})
        
        i += 1

    description = " ".join(description_lines)
    
    # Calculate hash
    # Use platform (category) from frontmatter as cluster if available
    # Otherwise fallback to directory name
    cluster = platform if platform else file_path.parent.name
    name = file_path.stem
    
    # Use cluster and name for hashing to match URL structure
    # The URL is /tldr/[cluster]/[name]
    hash_category = cluster
    
    full_hash = create_full_hash(hash_category, name)
    url_hash = get_8_bytes(full_hash)
    
    # Update path to match the new URL structure
    # This overrides the path from frontmatter which might be inconsistent
    path_url = f"{cluster}/{name}/"
    
    return {
        "url_hash": url_hash,
        "url": f"{hash_category}/{name}",
        "cluster": cluster,
        "name": name,
        "platform": platform,
        "title": title,
        "description": description,
        "more_info_url": more_info_url,
        "keywords": keywords,
        "features": features,
        "examples": json.dumps(examples),
        "raw_content": content,
        "path": path_url
    }


def process_all_files(conn: sqlite3.Connection) -> None:
    """Walk through DATA_DIR and populate the database."""
    cur = conn.cursor()
    
    # Prepare SQL
    insert_sql = """
    INSERT OR REPLACE INTO url_lookup (
        url_hash, url, cluster, name, platform, title, description, more_info_url,
        keywords, features, examples, raw_content, path
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    """
    
    files = list(DATA_DIR.glob("**/*.md"))
    print(f"Found {len(files)} markdown files.")
    
    batch = []
    count = 0
    
    for file_path in files:
        data = parse_tldr_file(file_path)
        if data:
            batch.append((
                data["url_hash"],
                data["url"],
                data["cluster"],
                data["name"],
                data["platform"],
                data["title"],
                data["description"],
                data["more_info_url"],
                data["keywords"],
                data["features"],
                data["examples"],
                data["raw_content"],
                data["path"]
            ))
            
            if len(batch) >= 100:
                cur.executemany(insert_sql, batch)
                conn.commit()
                count += len(batch)
                batch = []
                print(f"Processed {count} files...", end="\r")
    
    if batch:
        cur.executemany(insert_sql, batch)
        conn.commit()
        count += len(batch)
    
    print(f"\nSuccessfully inserted {count} pages.")


def populate_cluster_and_overview(conn: sqlite3.Connection) -> None:
    """Populate cluster and overview tables."""
    cur = conn.cursor()
    
    print("Populating cluster table...")
    cur.execute("DELETE FROM cluster")
    cur.execute("""
        INSERT INTO cluster (name, count, description)
        SELECT cluster, COUNT(*), '' 
        FROM url_lookup 
        GROUP BY cluster
    """)
    
    print("Populating overview table...")
    cur.execute("DELETE FROM overview")
    cur.execute("""
        INSERT INTO overview (id, total_count)
        SELECT 1, COUNT(*) FROM url_lookup
    """)
    
    conn.commit()


def main() -> None:
    # Ensure output directory exists
    DB_PATH.parent.mkdir(parents=True, exist_ok=True)
    
    # Remove existing DB to start fresh (optional, but good for full rebuilds)
    if DB_PATH.exists():
        DB_PATH.unlink()

    print(f"Creating database at {DB_PATH}")
    with sqlite3.connect(DB_PATH) as conn:
        ensure_schema(conn)
        process_all_files(conn)
        populate_cluster_and_overview(conn)
        
        # Verify
        cur = conn.cursor()
        cur.execute("SELECT total_count FROM overview")
        row = cur.fetchone()
        if row:
            print(f"Total pages in DB: {row[0]}")
        else:
            print("Total pages in DB: 0")

if __name__ == "__main__":
    main()

File: frontend/scripts/tldr/tldr_to_db.go
package main

import (
	"crypto/sha256"
	"database/sql"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io/fs"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/gomarkdown/markdown"
	"github.com/gomarkdown/markdown/html"
	"github.com/gomarkdown/markdown/parser"
	_ "github.com/mattn/go-sqlite3"
	"gopkg.in/yaml.v3"
)

// Paths
const (
	DataDir = "../../data/tldr"
	DbPath  = "../../db/all_dbs/tldr-db-v1.db"
)

// Structs matching DB schema
type Page struct {
	UrlHash     int64
	Url         string // Kept for reference
	Title       string
	Description string
	HtmlContent string
	Metadata    string // JSON (keywords, features)
}

type MainPage struct {
	Hash       int64
	Data       string // JSON
	TotalCount int
	Url        string
}

// Intermediate struct for processing
type ProcessedPage struct {
	UrlHash     int64
	Url         string
	Cluster     string
	Name        string
	Platform    string
	Title       string
	Description string
	Keywords    []string
	Features    []string
	HtmlContent string
	Path        string
}

type PageMetadata struct {
	Keywords []string `json:"keywords"`
	Features []string `json:"features"`
}

type Frontmatter struct {
	Title       string   `yaml:"title"`
	Description string   `yaml:"description"`
	Category    string   `yaml:"category"`
	Path        string   `yaml:"path"`
	Keywords    []string `yaml:"keywords"`
	Features    []string `yaml:"features"`
}

// --- Hashing Functions ---

func createFullHash(category, lastPath string) string {
	category = strings.ToLower(strings.TrimSpace(category))
	lastPath = strings.ToLower(strings.TrimSpace(lastPath))
	uniqueStr := fmt.Sprintf("%s/%s", category, lastPath)
	hash := sha256.Sum256([]byte(uniqueStr))
	return hex.EncodeToString(hash[:])
}

func get8Bytes(fullHash string) int64 {
	hexPart := fullHash[:16]
	bytesVal, err := hex.DecodeString(hexPart)
	if err != nil {
		log.Fatalf("Failed to decode hex: %v", err)
	}
	return int64(binary.BigEndian.Uint64(bytesVal))
}

func hashString(s string) string {
	hash := sha256.Sum256([]byte(s))
	return hex.EncodeToString(hash[:])
}

// --- Markdown Parsing ---

func parseTldrFile(path string) (*ProcessedPage, error) {
	contentBytes, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	content := string(contentBytes)

	parts := regexp.MustCompile(`(?m)^---\s*$`).Split(content, 3)
	if len(parts) < 3 {
		if strings.HasPrefix(content, "---") {
			parts = regexp.MustCompile(`(?m)^---\s*$`).Split(content, 3)
			if len(parts) >= 3 && parts[0] == "" {
			} else {
				return nil, nil
			}
		} else {
			return nil, nil
		}
	}

	frontmatterRaw := parts[1]
	markdownBody := strings.TrimSpace(parts[2])

	lines := strings.Split(markdownBody, "\n")
	if len(lines) > 0 && strings.HasPrefix(strings.TrimSpace(lines[0]), "# ") {
		markdownBody = strings.Join(lines[1:], "\n")
	}
	markdownBody = strings.TrimSpace(markdownBody)

	var fm Frontmatter
	if err := yaml.Unmarshal([]byte(frontmatterRaw), &fm); err != nil {
		log.Printf("Error parsing YAML in %s: %v", filepath.Base(path), err)
		return nil, nil
	}

	title := fm.Title
	description := fm.Description

	extensions := parser.CommonExtensions | parser.AutoHeadingIDs | parser.NoEmptyLineBeforeBlock
	p := parser.NewWithExtensions(extensions)
	doc := p.Parse([]byte(markdownBody))

	htmlFlags := html.CommonFlags | html.HrefTargetBlank
	opts := html.RendererOptions{Flags: htmlFlags}
	renderer := html.NewRenderer(opts)
	htmlBytes := markdown.Render(doc, renderer)
	htmlContent := string(htmlBytes)

	cluster := filepath.Base(filepath.Dir(path))
	name := strings.TrimSuffix(filepath.Base(path), filepath.Ext(path))
	fullHash := createFullHash(cluster, name)
	urlHash := get8Bytes(fullHash)
	pathUrl := fm.Path
	if pathUrl == "" {
		pathUrl = fmt.Sprintf("/freedevtools/tldr/%s/%s/", cluster, name)
	}
	if !strings.HasSuffix(pathUrl, "/") {
		pathUrl += "/"
	}

	return &ProcessedPage{
		UrlHash:     urlHash,
		Url:         pathUrl,
		Cluster:     cluster,
		Name:        name,
		Platform:    fm.Category,
		Title:       title,
		Description: description,
		Keywords:    fm.Keywords,
		Features:    fm.Features,
		HtmlContent: htmlContent,
		Path:        pathUrl,
	}, nil
}

// --- Database ---

func ensureSchema(db *sql.DB) error {
	// Drop tables if they exist to force schema update
	if _, err := db.Exec("DROP TABLE IF EXISTS pages"); err != nil {
		return err
	}
	if _, err := db.Exec("DROP TABLE IF EXISTS cluster"); err != nil {
		return err
	}
	if _, err := db.Exec("DROP TABLE IF EXISTS overview"); err != nil {
		return err
	}

	// Pages table - Simplified
	fmt.Println("Creating pages table...")
	_, err := db.Exec(`
		CREATE TABLE pages (
			url_hash INTEGER PRIMARY KEY,
			url TEXT NOT NULL, -- Kept for reference
			cluster_hash INTEGER NOT NULL,
			title TEXT DEFAULT '',
			description TEXT DEFAULT '',
			html_content TEXT DEFAULT '',
			metadata TEXT DEFAULT '{}' -- JSON (keywords, features)
		) WITHOUT ROWID;
	`)
	if err != nil {
		return fmt.Errorf("failed to create pages table: %w", err)
	}

	// Cluster table - Cluster Metadata (similar to emojis category)
	fmt.Println("Creating cluster table...")
	_, err = db.Exec(`
		CREATE TABLE cluster (
			hash INTEGER PRIMARY KEY,
			name TEXT NOT NULL,
			count INTEGER NOT NULL,
			preview_commands_json TEXT DEFAULT '[]' -- JSON list of preview commands
		) WITHOUT ROWID;
	`)
	if err != nil {
		return fmt.Errorf("failed to create cluster table: %w", err)
	}

	// Overview table
	fmt.Println("Creating overview table...")
	_, err = db.Exec(`
		CREATE TABLE overview (
			id INTEGER PRIMARY KEY CHECK(id = 1),
			total_count INTEGER NOT NULL,
			total_clusters INTEGER NOT NULL DEFAULT 0,
			total_pages INTEGER NOT NULL DEFAULT 0
		);
	`)
	if err != nil {
		return fmt.Errorf("failed to create overview table: %w", err)
	}
	return nil
}

func main() {
	start := time.Now()

	if err := os.MkdirAll(filepath.Dir(DbPath), 0755); err != nil {
		log.Fatal(err)
	}
	os.Remove(DbPath)

	db, err := sql.Open("sqlite3", DbPath)
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	if err := ensureSchema(db); err != nil {
		log.Fatal(err)
	}

	// 1. Parse all files into memory
	fmt.Printf("Scanning %s...\n", DataDir)
	var allPages []*ProcessedPage
	var allUrls []string

	err = filepath.WalkDir(DataDir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if !d.IsDir() && strings.HasSuffix(d.Name(), ".md") {
			page, err := parseTldrFile(path)
			if err != nil {
				log.Printf("Error parsing %s: %v", path, err)
				return nil
			}
			if page != nil {
				allPages = append(allPages, page)
			}
		}
		return nil
	})
	if err != nil {
		log.Fatal(err)
	}

	// 2. Insert Pages
	fmt.Println("Inserting pages...")
	tx, err := db.Begin()
	if err != nil {
		log.Fatal(err)
	}
	stmt, err := tx.Prepare("INSERT INTO pages (url_hash, url, cluster_hash, title, description, html_content, metadata) VALUES (?, ?, ?, ?, ?, ?, ?)")
	if err != nil {
		log.Fatal(err)
	}
	defer stmt.Close()

	for _, p := range allPages {
		meta := PageMetadata{
			Keywords: p.Keywords,
			Features: p.Features,
		}
		metaJson, _ := json.Marshal(meta)

		// Calculate cluster hash
		clusterHash := get8Bytes(hashString(p.Cluster))

		_, err = stmt.Exec(p.UrlHash, p.Url, clusterHash, p.Title, p.Description, p.HtmlContent, string(metaJson))
		if err != nil {
			log.Printf("Error inserting page %s: %v", p.Name, err)
		}
		allUrls = append(allUrls, p.Url)
	}
	tx.Commit()

	// 3. Generate Cluster (Cluster Metadata)
	fmt.Println("Generating cluster metadata...")
	pagesByCluster := make(map[string][]*ProcessedPage)
	for _, p := range allPages {
		pagesByCluster[p.Cluster] = append(pagesByCluster[p.Cluster], p)
	}

	tx, err = db.Begin()
	if err != nil {
		log.Fatal(err)
	}
	stmt, err = tx.Prepare("INSERT INTO cluster (hash, name, count, preview_commands_json) VALUES (?, ?, ?, ?)")
	if err != nil {
		log.Fatal(err)
	}

	for cluster, pages := range pagesByCluster {
		// Sort pages by name
		sort.Slice(pages, func(i, j int) bool {
			return pages[i].Name < pages[j].Name
		})

		totalCount := len(pages)

		// Get top 5 commands for preview
		var commandPreviews []map[string]string
		previewCount := 5
		if len(pages) < previewCount {
			previewCount = len(pages)
		}
		for k := 0; k < previewCount; k++ {
			commandPreviews = append(commandPreviews, map[string]string{
				"name": pages[k].Name,
				"url":  fmt.Sprintf("/freedevtools/tldr/%s/%s/", cluster, pages[k].Name),
			})
		}
		previewJson, _ := json.Marshal(commandPreviews)

		// Hash: cluster (e.g., common)
		hash := get8Bytes(hashString(cluster))

		_, err = stmt.Exec(hash, cluster, totalCount, string(previewJson))
		if err != nil {
			log.Printf("Error inserting cluster %s: %v", cluster, err)
		}
	}
	tx.Commit()

	// 4. Overview
	totalClusters := len(pagesByCluster)
	totalPages := len(allPages)
	if _, err := db.Exec("INSERT INTO overview (id, total_count, total_clusters, total_pages) VALUES (1, ?, ?, ?)", totalPages, totalClusters, totalPages); err != nil {
		log.Fatal(err)
	}

	elapsed := time.Since(start)
	fmt.Printf("Finished! Processed %d pages and %d URLs in %s.\n", len(allPages), len(allUrls), elapsed)
}
